File: 2pods/loadtest.go
package main

import (
    "context"
    "fmt"
    "log"
    "net/http"
    "os"
    "os/exec"
    "os/signal"
    "path/filepath"
    "sync"
    "syscall"
    "time"
)

const (
    namespace     = "web-observability"
    url           = "http://127.0.0.1:18080"
    pfLogFilename = "one-pod-web-port-forward.log"
)

type Stage struct {
    Concurrency int
    Duration    int
    Work        int
}

func main() {
    // 1. Setup Paths & Env Vars
    root, err := os.Getwd()
    if err != nil {
        log.Fatalf("Failed to get working directory: %v", err)
    }
    os.Setenv("KUBECONFIG", filepath.Join(root, ".kubeconfig"))

    tmpDir := os.Getenv("TMPDIR")
    if tmpDir == "" {
        tmpDir = "/tmp"
    }
    logPath := filepath.Join(tmpDir, pfLogFilename)

    // 2. Context for graceful shutdown (Replaces bash `trap cleanup EXIT INT TERM`)
    ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
    defer stop()

    // 3. Start Port-Forward
    logFile, err := os.Create(logPath)
    if err != nil {
        log.Fatalf("Failed to create log file: %v", err)
    }
    defer logFile.Close()

    // CommandContext automatically kills the process if ctx is canceled
    pfCmd := exec.CommandContext(ctx, "kubectl", "--namespace", namespace, "port-forward", "service/web", "18080:8080")
    pfCmd.Stdout = logFile
    pfCmd.Stderr = logFile

    if err := pfCmd.Start(); err != nil {
        log.Fatalf("Failed to start port-forward: %v", err)
    }
    fmt.Printf("Started port-forward (PID %d), logging to %s\n", pfCmd.Process.Pid, logPath)

    // 4. Readiness Check (Replaces the 30-iteration bash loop)
    fmt.Println("Waiting for port-forward to become ready...")
    client := &http.Client{Timeout: 2 * time.Second}
    ready := false
    for i := 0; i < 30; i++ {
        resp, err := client.Get(url + "/healthz")
        if err == nil && resp.StatusCode == http.StatusOK {
            resp.Body.Close()
            ready = true
            break
        }
        if resp != nil {
            resp.Body.Close()
        }
        time.Sleep(1 * time.Second)
    }

    if !ready {
        log.Fatalf("web port-forward did not become ready; see %s", logPath)
    }
    fmt.Println("Port-forward is ready!")

    // 5. Start CPU Monitor Goroutine (Replaces bash background function)
    go monitorCPU(ctx)

    // 6. Define Test Stages
    singleRamp := []Stage{
        {1, 10, 250000},
        {5, 15, 1000000},
        {15, 25, 3000000},
        {30, 30, 5000000},

File: 2pods/loadtest.py
#!/usr/bin/env python3
import os
import sys
import time
import signal
import subprocess
import threading
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor, as_completed

import requests

# --- Configuration & Setup ---
ROOT = Path(__file__).resolve().parent.parent
os.environ["KUBECONFIG"] = str(ROOT / ".kubeconfig")

NAMESPACE = "web-observability"
URL = "http://127.0.0.1:18080"
PORT_FORWARD_LOG = Path(os.environ.get("TMPDIR", "/tmp")) / "one-pod-web-port-forward.log"

# Global state for cleanup
pf_proc = None
log_file = None
stop_monitor_event = threading.Event()

def cleanup():
    """Mirrors the bash trap cleanup function."""
    stop_monitor_event.set()  # Signal the monitor thread to stop
    
    if pf_proc:
        pf_proc.terminate()
        try:
            pf_proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            pf_proc.kill()
            
    if log_file and not log_file.closed:
        log_file.close()

def signal_handler(sig, frame):
    cleanup()
    sys.exit(0)

# Register signal handlers (equivalent to `trap cleanup EXIT INT TERM`)
signal.signal(signal.SIGINT, signal_handler)
signal.signal(signal.SIGTERM, signal_handler)

def monitor_cpu():
    """Mirrors the bash monitor_cpu function."""
    while not stop_monitor_event.is_set():
        print()
        print(f"CPU sample {time.strftime('%H:%M:%S')}")
        try:
            subprocess.run(
                ["kubectl", "top", "pods", "--namespace", NAMESPACE, 
                 "--selector", "app.kubernetes.io/name=web"],
                check=False,
                stdout=sys.stdout,
                stderr=subprocess.DEVNULL
            )
        except FileNotFoundError:
            print("CPU metrics are warming up (kubectl not found)")
            
        # Use event.wait() instead of time.sleep() so it reacts instantly to cleanup
        stop_monitor_event.wait(5)

def run_stage(concurrency: int, duration: int, work: int):
    """Mirrors the bash run_stage function using a thread pool."""
    print(f"\nStage: concurrency={concurrency} duration={duration}s work={work}")
    deadline = time.time() + duration

    def worker():
        while time.time() < deadline:
            try:
                # Mimics: curl --silent --show-error --fail --max-time 30 --output /dev/null
                requests.get(f"{URL}/?work={work}", timeout=30)
            except requests.exceptions.RequestException:
                # Catch timeouts, connection errors, etc., but keep looping until deadline
                pass

    # Spawn threads instead of background bash processes
    with ThreadPoolExecutor(max_workers=concurrency) as executor:
        futures = [executor.submit(worker) for _ in range(concurrency)]
        for future in as_completed(futures):
            # Re-raise any unexpected thread exceptions
            future.result()

# --- Main Execution ---
try:
    # 1. Start port-forward in the background
    log_file = open(PORT_FORWARD_LOG, "w")
    pf_proc = subprocess.Popen(
        ["kubectl", "--namespace", NAMESPACE, "port-forward", "service/web", "18080:8080"],
        stdout=log_file,
        stderr=subprocess.STDOUT
    )

    # 2. Wait for readiness (equivalent to the 30-iteration loop)
    print("Waiting for port-forward to become ready...")
    for _ in range(30):
        try:
            if requests.get(f"{URL}/healthz", timeout=2).ok:
                break
        except requests.exceptions.RequestException:
            pass
        time.sleep(1)
    else:
        # Executed if the loop didn't 'break'
